Sources/NIOHTTP1/NIOHTTPClientUpgradeHandler.swift

```diff
@@ -0,0 +1,434 @@
+            context.fireErrorCaught(NIOHTTPClientUpgradeError.responseProtocolNotFound)
+            self.notUpgrading(context: context, data: responsePart)
+            return
+        }
+
+        // This is an upgrade response.
+        // This may take a while, so while we're waiting more data can come in.
+        self.upgradeState = .awaitingUpgrader
+        
+        self.handleUpgrade(context: context, upgradeResponse: response, receivedProtocols: acceptedProtocols).whenSuccess { callback in
+
+            if let callback = callback {
+                self.gotUpgrader(upgrader: callback)
+            } else {
+                self.notUpgrading(context: context, data: responsePart)
+            }
+        }
+    }
+    
+    private func handleUpgrade(context: ChannelHandlerContext, upgradeResponse response: HTTPResponseHead, receivedProtocols: [String]) -> EventLoopFuture<(() -> Void)?> {
+        
+        let allHeaderNames = Set(response.headers.map { $0.name.lowercased() })
+        
+        // We now set off a chain of Futures to try to find a protocol upgrade.
+        // While this is blocking, we need to buffer inbound data.
+        let protocolIterator = receivedProtocols.makeIterator()
+        
+        return self.handleUpgradeForProtocol(context: context, protocolIterator: protocolIterator, response: response, allHeaderNames: allHeaderNames)
+    }
+    
+    /// Attempt to upgrade a single protocol.
+    private func handleUpgradeForProtocol(context: ChannelHandlerContext, protocolIterator: Array<String>.Iterator, response: HTTPResponseHead, allHeaderNames: Set<String>) -> EventLoopFuture<(() -> Void)?> {
+        
+        var protocolIterator = protocolIterator
+        guard let lowercaseProtoName = protocolIterator.next()?.lowercased() else {
+            // There are no upgrade protocols returned.
+            context.fireErrorCaught(NIOHTTPClientUpgradeError.responseProtocolNotFound)
+            return context.eventLoop.makeSucceededFuture(nil)
+        }
+
+        let matchingUpgrader = self.upgraders
+            .first(where: { $0.supportedProtocol.lowercased() == lowercaseProtoName })
+        
+        guard let upgrader = matchingUpgrader else {
+            // There is no upgrader for this protocol.
+            context.fireErrorCaught(NIOHTTPClientUpgradeError.responseProtocolNotFound)
+            return context.eventLoop.makeSucceededFuture(nil)
+        }
+        
+        guard upgrader.shouldAllowUpgrade(upgradeResponse: response) else {
+            // The upgrader says no.
+            context.fireErrorCaught(NIOHTTPClientUpgradeError.upgraderDeniedUpgrade)
+            return context.eventLoop.makeSucceededFuture(nil)
+        }
+        
+        let upgradeFunction = self.performUpgrade(context: context, upgrader: upgrader, response: response)
+        
+        return context.eventLoop.makeSucceededFuture(upgradeFunction)
+        .flatMapError { error in
```

Comment:
This `flatMapError` cannot possibly be called as you're chaining it from a `succeededFuture`. What's it for?

Reply:
I think I just forgot to remove it from a previous implementation. Done now. Thanks.